DistroLauncher/named_mutex.h

```diff
@@ -0,0 +1,159 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
+
+#include "stdafx.h"
+
+/*
+ *  Abstract
+ *  --------
+ *  A RAII class to deal with Windows inter-process synchronization.
+ *
+ *  Usage
+ *  --------
+ *  Best is to create a single NamedMutex as a static variable. If you believe
+ *  it's not going to be used in most executions, then use
+ *  ```
+ *  static my_mutex = NamedMutex("unique_identifyier", true);
+ *  ```
+ *  to enable lazy initialization. Otherwise, use
+ *  ```
+ *  static my_mutex = NamedMutex("unique_identifyier");
+ *  ```
+ *
+ *  In order to acquire a lock, you have two options.
+ *  Use the first one unless you've got a good reason not to.
+ *   1. Use a lambda wrapper:
+ *      ```
+ *      DWORD response = my_mutex.Run([](){
+ *          // do stuff
+ *      });
+ *      if(response != 0) {
+ *          throw "Oh, no!";
+ *      }
+ *      ```
+ *
+ *   2. Use a scope guard:
+ *      ```
+ *      NamedMutex::Guard scope_guard = my_mutex.get();
+ *      if(scope_guard.ok()) {
+ *          // do stuff
+ *      } else {
+ *          throw "Oh, no!";
+ *      }
+ *      ```
+ *      Be careful! If your program throws and it's not caught,
+ *      your mutex may not be released. Option 1 takes care of
+ *      this automatically, hence why it's recommended.
+ */
+class NamedMutex
+{
+  public:
+    NamedMutex() = delete;
+    NamedMutex(NamedMutex&) = delete;
+    NamedMutex(NamedMutex&&) = delete;
+
+    /// Some mutexes are almost never used. This makes for an easy optimization by not initializing them unless needed.
+    NamedMutex(std::wstring name, bool lazy_init = false) :
+        mutex_handle(nullptr), mutex_name(L"WSL_" + DistributionInfo::Name + L"_" + name)
+    {
+        if (!lazy_init) {
+            create();
+        }
+    }
+
+    ~NamedMutex()
+    {
+        destroy();
+    }
+
+    class Guard
+    {
+      public:
+        constexpr Guard() noexcept : parent_(nullptr), response_(0)
+        { }
+
+        Guard(NamedMutex& parent) noexcept : parent_(&parent), response_(parent.wait_and_acquire())
+        { }
+
+        Guard(Guard& other) = delete;
+        Guard(Guard&& other) noexcept : Guard()
+        {
+            *this = Guard(std::move(other));
+        }
+
+        ~Guard() noexcept
+        {
+            if (ok()) {
+                parent_->release();
+            }
+        }
+
+        Guard& operator=(Guard&& other) noexcept
+        {
```

Comment:
This seem suspecious to me: you're modifying both Guards by swapping their guts on an assignment operator call? If that's really what you want, it'd be better defining a free function called swap in the same namespace of the class.

Reply:
This is in case you want to pass the guard around. For instance:

```c++
struct Sudo {
NamedMutex::Guard sudo_guard (nullptr);

Sudo() {
   static auto root_user_mutex = NamedMutex("sudo", true);
   auto guard = root_user_mutex.lock();

   auto succeeded = SetUserAsRoot();
   
   if(succeeded) {
      this->sudo_guard = std::move(guard); // extending lifetime
   }
}

~Sudo() {
   ResetUser();
}

};

```

This way you can extend the guard lifetime outside its scope.

Reply:
Real example here:

https://github.com/ubuntu/WSL/blob/8f9f6a60a4f7af49245eb72c5246414ed778f9a8/DistroLauncher/sudo.h#L111-L133